### PR TITLE
Enforce `CADisableMinimumFrameDurationOnPhone`

### PR DESCRIPTION
**Pull request by elijah-semyonov**
## Proposed Changes

By default, crash applications which don't contain a proper `CADisableMinimumFrameDurationOnPhone` entry in the Info.plist, when running on an iPhone.

Implementation for https://youtrack.jetbrains.com/issue/CMP-5643/Make-CADisableMinimumFrameDurationOnPhone-a-requirement-to-run-Compose-Multiplatform-on-iOS

### iOS - Breaking Changes
- The app will crash by default, if `CADisableMinimumFrameDurationOnPhone` is not set to true in `Info.plist`. Use newly added `ComposeUIViewControllerConfiguration.enforceStrictPlistSanityCheck` to opt-out of this behavior.

### CODE REVIEW

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/PlistSanityCheck.uikit.kt [hunk 1]
```diff
@@ -37,7 +37,15 @@ internal object PlistSanityCheck {
                     .objectForInfoDictionaryKey("CADisableMinimumFrameDurationOnPhone") as? NSNumber
 
                 if (entry?.boolValue != true) {
-                    println("WARNING: `Info.plist` doesn't have a valid `CADisableMinimumFrameDurationOnPhone` entry. Framerate will be restricted to 60hz on iPhones. To support high frequency rendering on iPhones, add `<key>CADisableMinimumFrameDurationOnPhone</key><true/>` entry to `Info.plist`.")
+                    if (enforceStrictPlistSanityCheck) {
+                        val message = "Error: `Info.plist` doesn't have a valid `CADisableMinimumFrameDurationOnPhone` entry. This will result in an inadequate performance on devices with high refresh rate. Add `<key>CADisableMinimumFrameDurationOnPhone</key><true/>` entry to `Info.plist` to fix this error. Or set `ComposeUIViewControllerConfiguration.enforceStrictPlistSanityCheck` to false, if it's intended."
```
Comment (terrakok) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/PlistSanityCheck.uikit.kt]: There is no such thing as `Info.plist` anymore.
https://stackoverflow.com/questions/67896404/where-is-info-plist-in-xcode-13-missing-not-inside-project-navigator

Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/PlistSanityCheck.uikit.kt]: Modified the message to be more explicit about this scenario

**Review by terrakok — APPROVED** [review]

**Comment by igordmn** [comment]
> iOS - Features

Please, change it to `Breaking Changes - iOS`

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/ComposeUIViewControllerConfiguration.uikit.kt [hunk 2]
```diff
@@ -55,6 +55,13 @@ class ComposeUIViewControllerConfiguration {
      */
     @ExperimentalComposeApi
     var opaque: Boolean = true
+
+    /**
+     * A boolean flag to enable or disable the strict sanity check for the `Info.plist` file.
+     * In case some of the required keys are missing, a warning will be printed to the console.
```
Comment (igordmn) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/ComposeUIViewControllerConfiguration.uikit.kt]:
```suggestion
```

If users explicitly opt-out, they don't want to see a warning. Either an error or nothing are enough.

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/PlistSanityCheck.uikit.kt [hunk 3]
```diff
@@ -37,7 +37,21 @@ internal object PlistSanityCheck {
...
+                            Error: `Info.plist` doesn't have a valid `CADisableMinimumFrameDurationOnPhone` entry. 
+                            This will result in an inadequate performance on devices with high refresh rate. 
+                            Add `<key>CADisableMinimumFrameDurationOnPhone</key><true/>` entry to `Info.plist` to fix this error. 
+                            If you don't have a separate plist file, add the entry to the target from within Xcode: Project -> Targets -> Info -> Custom iOS Target Properties.                            
+                            Or set `ComposeUIViewControllerConfiguration.enforceStrictPlistSanityCheck` to false, if it's intended.                            
+                        """.trimIndent()
+                        println(message)
+                        terminateWithUnhandledException(
+                            IllegalStateException(message)
+                        )
```
Comment (igordmn) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/PlistSanityCheck.uikit.kt]:
```suggestion
                        error(message)
```

`terminateWithUnhandledException` is native only and `ExperimentalNativeApi`. `error` on the other hand is common and established way to tell user that is something wrong.

Unless I don't know something about `terminateWithUnhandledException`, let's use `error`

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/PlistSanityCheck.uikit.kt [hunk 4]
```diff
@@ -37,7 +37,21 @@ internal object PlistSanityCheck {
...
 
                 if (entry?.boolValue != true) {
-                    println("WARNING: `Info.plist` doesn't have a valid `CADisableMinimumFrameDurationOnPhone` entry. Framerate will be restricted to 60hz on iPhones. To support high frequency rendering on iPhones, add `<key>CADisableMinimumFrameDurationOnPhone</key><true/>` entry to `Info.plist`.")
+                    if (enforceStrictPlistSanityCheck) {
+                        val message = """
+                            Error: `Info.plist` doesn't have a valid `CADisableMinimumFrameDurationOnPhone` entry. 
+                            This will result in an inadequate performance on devices with high refresh rate. 
+                            Add `<key>CADisableMinimumFrameDurationOnPhone</key><true/>` entry to `Info.plist` to fix this error. 
+                            If you don't have a separate plist file, add the entry to the target from within Xcode: Project -> Targets -> Info -> Custom iOS Target Properties.                            
+                            Or set `ComposeUIViewControllerConfiguration.enforceStrictPlistSanityCheck` to false, if it's intended.
```
Comment (igordmn) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/PlistSanityCheck.uikit.kt]:
```suggestion
                            Or set `ComposeUIViewController(configure = { enforceStrictPlistSanityCheck = false }) { .. }`, if it's intended.                            
```

Users don't see `ComposeUIViewControllerConfiguration` unless they go to the sources.

**Review by igordmn — CHANGES_REQUESTED** [review]

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/ComposeUIViewControllerConfiguration.uikit.kt (hunk 2 above)
Comment (igordmn) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/ComposeUIViewControllerConfiguration.uikit.kt]: The warning ("println") is still there. My suggestion was to remove it, not only the KDoc.

File: compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt [hunk 5]
```diff
@@ -81,7 +81,12 @@ internal class UIKitInstrumentedTest {
     ) {
         var onDraw = false
         composeContainer = ComposeContainer(
-            configuration = ComposeUIViewControllerConfiguration().apply(configure),
+            configuration = ComposeUIViewControllerConfiguration().apply {
+                // TODO: setup proper plist for instrumented test
+                //   https://youtrack.jetbrains.com/issue/CMP-5706/iOS-setup-proper-plist-in-instrumented-tests
+                enforceStrictPlistSanityCheck = false
```
Comment (igordmn) [inline compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:86]: Will it mean that user UI tests also fail? Could you check?

File: compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt [hunk 6]
```diff
@@ -81,7 +81,11 @@ internal class UIKitInstrumentedTest {
     ) {
         var onDraw = false
```
Comment (igordmn) [inline compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:82]: > iOS - Features

Please, change it to Breaking Changes - iOS

(moved from https://github.com/JetBrains/compose-multiplatform-core/pull/1451#issuecomment-2230485628)

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/ComposeUIViewControllerConfiguration.uikit.kt (hunk 2 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/ComposeUIViewControllerConfiguration.uikit.kt]: Got it

File: compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt (hunk 5 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:86]: I've fixed it

File: compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt (hunk 6 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:82]: Done

**Comment by elijah-semyonov** [comment]
Relaxed the check a change to crash only on iPhone. Apparently, iPad is not affected by this flag.

**Comment by igordmn** [comment]
> Relaxed the check a change to crash only on iPhone. Apparently, iPad is not affected by this flag.

Better to reenable it:
- devs who check on iPad can be surprised when run it on iPhone
- if we checked only one device, we can't be sure that this always happen

File: compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/PlistSanityCheck.uikit.kt [hunk 7]
```diff
@@ -32,12 +32,22 @@ internal object PlistSanityCheck {
...
+                if (displayLinkEntry?.boolValue != true) {
+                    val message = """
+                            Error: `Info.plist` doesn't have a valid `CADisableMinimumFrameDurationOnPhone` entry. 
+                            This will result in an inadequate performance on iPhones with high refresh rate. 
+                            Add `<key>CADisableMinimumFrameDurationOnPhone</key><true/>` entry to `Info.plist` to fix this error. 
+                            If you don't have a separate plist file, add the entry to the target from within Xcode: Project -> Targets -> Info -> Custom iOS Target Properties.                            
+                            Or set `ComposeUIViewController(configure = { enforceStrictPlistSanityCheck = false }) { .. }`, if it's intended.                            
+                        """.trimIndent()
+                    println(message)
```
Comment (igordmn) [inline compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/uikit/PlistSanityCheck.uikit.kt]:
```suggestion
```
It seems `error` will print the exception itself?

File: compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt (hunk 5 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:86]: Yes, it's a breaking change.

Comment (igordmn) [reply compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:86]: What I wonder, if it fail in these cases:
1. when user uses our `ui-test` API. Does it use `ComposeUIViewController` inside?
2. when user just uses `ComposeUIViewController` in the test. Does it uses the same plist as for the main application?

If we don't force users to specify `enforceStrictPlistSanityCheck = false` in their tests - it is good. If we force - it isn't, and we should weight all pros/cons.

Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:86]: Well, if we don't make this flag `true` by default, it kills the whole purpose of this venture. I doubt someone is using ui-test API yet, it was just brought.

Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:86]: Oh... I've just realised, that our test api on iOS is `internal`, so no blocker here?

Comment (igordmn) [reply compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:86]: `UIKitInstrumentedTest` is internal, yes, no blocker with it. But what is with the others?

Comment (igordmn) [reply compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:86]: https://www.jetbrains.com/help/kotlin-multiplatform-dev/compose-test.html#writing-and-running-tests-with-compose-multiplatform

Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:86]: We don't have other usages. If you mean other people rolling their own testing and getting a crash if plist is wrong - then it's by design. They can specifically opt out of this behavior if they don't need it for test. That's the whole rationale behind this PR.

Comment (elijah-semyonov) [reply compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:86]: It's completely normal that manifest inconsistencies cause a crash in runtime on iOS - a good example of it is lack of privacy description in plist, whenever a user attempts to access privacy-protected API.

**Review by igordmn — APPROVED** [review]

**Comment by elijah-semyonov** [comment]
Oh oh... still not merged

File: compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt (hunk 5 above)
Comment (ASalavei) [reply compose/ui/ui/src/uikitInstrumentedTest/kotlin/android/compose/ui/test/UIKitInstrumentedTest.kt:86]: It looks like for now instrumented tests aren't use plist file, and considering the fact we're not running tests on a real device, I would recommend to keep as it is and remove TODO part.

**Comment by elijah-semyonov** [comment]
Removed TODO

**Pull request merged**